### PR TITLE
Feature.l7switching

### PR DESCRIPTION
**Pull request by pjbreaux**
@jlongstaf and @szakeri 

#### What issues does this address?
Fixes #<issueid>
WIP #393

#### What's this change do?
Tempest scenario tests need to be added for l7policies and rules

#### Where should the reviewer start?
Start at the new tests for l7policies and rules.

#### Any background context?
We need to add tempest scenario tests for l7policies and l7 rules. These are needed for traffic validation after policies are deployed

### CODE REVIEW

File: f5lbaasdriver/test/tempest/tests/scenario/test_l7policies_and_rules.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    def _reject_args(self):
+        self.reject_args = {
+            'listener_id': self.listener.get('id'),
+            'admin_state_up': 'true', 'action': 'REJECT'}
+
+    def test_policy_reject_header_starts_with(self):
+        '''Reject traffic when header value starts with value.'''
+
+        self.l7policy = self._create_l7policy(**self.reject_args)
+        rule_args = {'type': 'HEADER', 'compare_type': 'CONTAINS',
```
Comment (szakeri) [inline f5lbaasdriver/test/tempest/tests/scenario/test_l7policies_and_rules.py]: Paul, the test name indicate that it wants to check compare type of "starts with", but in the rule args, compare_type is set to "CONTAINS"

File: f5lbaasdriver/test/tempest/tests/scenario/test_l7policies_and_rules.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        self._delete_l7policy(self.l7policy.get('id'))
+        self._wait_for_load_balancer_status(self.load_balancer.get('id'))
+        self._run_traffic('server1', headers={'X-HEADER': 'test'})
+        self._remove_members()
+
+    def test_policy_reject_header_starts_with_issue_393(self):
+        '''Test should fail until pool delete also deletes nodes on device.'''
+
+        self.l7policy = self._create_l7policy(**self.reject_args)
+        rule_args = {'type': 'HEADER', 'compare_type': 'CONTAINS',
```
Comment (szakeri) [inline f5lbaasdriver/test/tempest/tests/scenario/test_l7policies_and_rules.py]: Same here

File: f5lbaasdriver/test/tempest/tests/scenario/test_l7policies_and_rules.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        self.members_client.delete_member(
+            self.detached_pool.get('id'),
+            self.members['secondary'].get('id')
+        )
+        self._wait_for_load_balancer_status(self.load_balancer.get('id'))
+
+    def _run_traffic(
+            self, expected_server, headers=None, cookies=None,
+            expected_status=200):
+        print('##################')
```
Comment (szakeri) [inline f5lbaasdriver/test/tempest/tests/scenario/test_l7policies_and_rules.py]: Are these print statements for debugging?

Comment (pjbreaux) [inline f5lbaasdriver/test/tempest/tests/scenario/test_l7policies_and_rules.py]: Yes, I'd like to leave them in for the time being, so we can sanity check the expectations against the responses.

**Comment by pjbreaux** [comment]
Okay, created most of the tests we'll need here. All but one passes. The one that fails is for issue #393

```
root@170444c6cac8:~/f5-openstack-lbaasv2-driver# py.test f5lbaasdriver/test/tempest/tests/scenario/test_l7policies_and_rules.py
<class 'f5.bigip.BigIP'>
========================================= test session starts ==========================================
platform linux2 -- Python 2.7.12, pytest-2.9.1, py-1.4.31, pluggy-0.3.1
rootdir: /root/f5-openstack-lbaasv2-driver, inifile:
plugins: symbols-0.1.0a3, cov-2.4.0, f5-openstack-test-0.2.0
collected 11 items

f5lbaasdriver/test/tempest/tests/scenario/test_l7policies_and_rules.py ...........

===================================== 11 passed in 880.49 seconds ======================================
root@170444c6cac8:~/f5-openstack-lbaasv2-driver#
```

**Comment by szakeri** [comment]
Looks good.

**Comment by pjbreaux** [comment]
Added a few new tests. One important one is performing a policy reorder (via update) and a rule change (also via update). Rerunning all tests now and will post the results here.

**Comment by pjbreaux** [comment]
Just reran these and got the expected results:

```
root@e19a8e02b968:~/f5-openstack-lbaasv2-driver/f5lbaasdriver/test/tempest/tests/scenario# py.test -v test_l7policies_and_rules.py
<class 'f5.bigip.BigIP'>
==================================================================== test session starts =====================================================================
platform linux2 -- Python 2.7.12, pytest-3.0.4, py-1.4.31, pluggy-0.4.0 -- /usr/bin/python
cachedir: ../../../../../.cache
rootdir: /root/f5-openstack-lbaasv2-driver, inifile:
plugins: symbols-0.1.0a3, cov-2.4.0, f5-openstack-test-0.2.0
collected 13 items

test_l7policies_and_rules.py::TestL7BasicRedirectToPool::test_policy_deployment PASSED
test_l7policies_and_rules.py::TestL7BasicRedirectToPool::test_policy_redirect_pool_cookie_contains PASSED
test_l7policies_and_rules.py::TestL7BasicRedirectToPool::test_policy_redirect_pool_cookie_not_contains PASSED
test_l7policies_and_rules.py::TestL7BasicRedirectToPool::test_policy_redirect_pool_header_equal_to PASSED
test_l7policies_and_rules.py::TestL7BasicRedirectToPool::test_policy_redirect_pool_hostname_starts_with PASSED
test_l7policies_and_rules.py::TestL7BasicRedirectToUrl::test_policy_redirect_url_file_type_fails_bad_compare_type PASSED
test_l7policies_and_rules.py::TestL7BasicRedirectToUrl::test_policy_redirect_url_file_type_not_contains PASSED
test_l7policies_and_rules.py::TestL7BasicRedirectToUrl::test_policy_redirect_url_hostname_ends_with PASSED
test_l7policies_and_rules.py::TestL7BasicRedirectToUrl::test_policy_redirect_url_path_starts_with PASSED
test_l7policies_and_rules.py::TestL7BasicReject::test_policy_reject_file_type_equal_to PASSED
test_l7policies_and_rules.py::TestL7BasicReject::test_policy_reject_header_starts_with PASSED
test_l7policies_and_rules.py::TestL7BasicUpdate::test_policy_reorder_and_rule_update PASSED
test_l7policies_and_rules.py::TestL7BasicUpdate::test_policy_sol17095_deployment_operand_match FAILED
test_l7policies_and_rules.py::TestL7BasicUpdate::test_policy_sol17095_deployment_operand_match ERRO
```

**Comment by pjbreaux** [comment]
Okay, ran again and still got the passes I expect:

```
root@e19a8e02b968:~/f5-openstack-lbaasv2-driver/f5lbaasdriver/test/tempest/tests/scenario# py.test -v test_l7policies_and_rules.py
<class 'f5.bigip.BigIP'>
================================================ test session starts =================================================
platform linux2 -- Python 2.7.12, pytest-3.0.4, py-1.4.31, pluggy-0.4.0 -- /usr/bin/python
cachedir: ../../../../../.cache
rootdir: /root/f5-openstack-lbaasv2-driver, inifile:
plugins: symbols-0.1.0a3, cov-2.4.0, f5-openstack-test-0.2.0
collected 13 items

test_l7policies_and_rules.py::TestL7BasicRedirectToPool::test_policy_deployment PASSED
test_l7policies_and_rules.py::TestL7BasicRedirectToPool::test_policy_redirect_pool_cookie_contains PASSED
test_l7policies_and_rules.py::TestL7BasicRedirectToPool::test_policy_redirect_pool_cookie_not_contains PASSED
test_l7policies_and_rules.py::TestL7BasicRedirectToPool::test_policy_redirect_pool_header_equal_to PASSED
test_l7policies_and_rules.py::TestL7BasicRedirectToPool::test_policy_redirect_pool_hostname_starts_with PASSED
test_l7policies_and_rules.py::TestL7BasicRedirectToUrl::test_policy_redirect_url_file_type_fails_bad_compare_type PASSED
test_l7policies_and_rules.py::TestL7BasicRedirectToUrl::test_policy_redirect_url_file_type_not_contains PASSED
test_l7policies_and_rules.py::TestL7BasicRedirectToUrl::test_policy_redirect_url_hostname_ends_with PASSED
test_l7policies_and_rules.py::TestL7BasicRedirectToUrl::test_policy_redirect_url_path_starts_with PASSED
test_l7policies_and_rules.py::TestL7BasicReject::test_policy_reject_file_type_equal_to PASSED
test_l7policies_and_rules.py::TestL7BasicReject::test_policy_reject_header_starts_with PASSED
test_l7policies_and_rules.py::TestL7BasicUpdate::test_policy_deployment_operand_match FAILED
test_l7policies_and_rules.py::TestL7BasicUpdate::test_policy_reorder_and_rule_update PASSED
```

**Comment by pjbreaux** [comment]
I'm ready to merge when anyone else gives the okay.

**Pull request merged**